packages/core/src/app/pg-core/db/chain-db.ts

```diff
@@ -1,24 +1,523 @@
+   * @param blockNumber Block number to query.
+   * @returns the hash of the given block.
+   */
+  public async getBlockHash(blockNumber: number): Promise<string> {
+    const key = KEYS.BLOCKS.encode([blockNumber])
+    const value = await this.db.get(key)
+    return value.toString()
+  }
+
+  /**
+   * Queries a specific stored Merkle tree node. Attempts to compute the node
+   * from its children if the node doesn't exist.
+   * @param blockNumber Block number to pull the node from.
+   * @param nodeIndex Index of the specific node to query.
+   * @returns the node or `null` if the node cannot be found.
+   */
+  public async getMerkleTreeNode(
+    blockNumber: number,
+    nodeIndex: number
+  ): Promise<string> {
+    // Reject attempts to query invalid nodes.
+    const treeHeight = await this.getBlockTreeHeight(blockNumber)
+    if (nodeIndex < 0 || nodeIndex >= treeHeight) {
+      throw new Error('Invalid node index.')
+    }
+
+    const key = KEYS.TREE_NODES.encode([blockNumber, nodeIndex])
+    const value = await this.db.get(key)
+
+    // We have this node, return it.
+    if (value !== null) {
+      return value.toString()
+    }
+
+    // We don't have the node, but we might be able to compute the it from its
+    // children. Recursively find each child and see if we can create the node.
+
+    // We're at a leaf node so we can't go any deeper.
+    if (nodeIndex > 2 ** treeHeight) {
+      return null
+    }
+
+    // Compute left and right children indices.
+    const leftChildIndex = getLeftChildIndex(nodeIndex)
+    const rightChildIndex = getRightChildIndex(nodeIndex)
+
+    // Recursively pull the left and right children.
+    const leftChild = await this.getMerkleTreeNode(blockNumber, leftChildIndex)
+    const rightChild = await this.getMerkleTreeNode(
+      blockNumber,
+      rightChildIndex
+    )
+
+    // We can't create a parent if either child is node.
+    if (leftChild === null || rightChild === null) {
+      return null
+    }
+
+    // We have both children so we can compute the parent!
+    return computeParent(leftChild, rightChild)
+  }
+
+  /**
+   * Adds a Merkle tree node for a specific block. Won't add the node if it
+   * already exists in the database or can be computed from other nodes that
+   * exist.
+   * @param blockNumber Block to add the node for.
+   * @param nodeIndex Index of the node to add.
+   * @param nodeHash Hash of the node to add.
+   */
+  public async addMerkleTreeNode(
+    blockNumber: number,
+    nodeIndex: number,
+    nodeHash: string
+  ): Promise<void> {
+    // Don't do anything if we already have the node.
+    if ((await this.getMerkleTreeNode(blockNumber, nodeIndex)) !== null) {
+      return
+    }
+
+    // Smart parent checks. We don't need to store the parent as long as we
+    // have this node's sibling. Parent can simply be re-generated later on.
+    // Skip if this is the root node because it doesn't have a sibling.
+    if (nodeIndex !== 0) {
```

Comment:
If we're adding a node that has a sibling, then we no longer need to keep the node's parent since we can generate it from the two children.